### PR TITLE
Added Synonyms Support

### PR DESCRIPTION
**Pull request by MeirShpilraien**
It is now possible to add synonyms group and use synonyms on search queries.
Look at Synonyms.md for more information.

### CODE REVIEW

File: src/document.c [hunk 1]
```diff
@@ -135,6 +135,12 @@ RSAddDocumentCtx *NewAddDocumentCtx(IndexSpec *sp, Document *b, const char **err
     aCtx->fwIdx = NewForwardIndex(&aCtx->doc, sp->flags);
   }
 
+  if (sp->smap){
+    aCtx->fwIdx->smap = SynonymMap_GetReadOnlyCopy(sp->smap);
+  }else{
```
Comment (mnunberg) [inline src/document.c]: spacing again here.. i use vscode for formatting and you should too; makes things easier. do you have a preferred editor?

**Comment by MeirShpilraien** [comment]
@mnunberg the thread safe is not an issue. I am passing a read only copy of the synonyms data structure to the ForwardIndex. The read only copy is managed by ref count.

File: src/document.c (hunk 1 above)
Comment (MeirShpilraien) [reply src/document.c]: Just installed vscode, I am learning how to use it.

**Comment by mnunberg** [comment]
I see now, you have a read only copy which is detached from the actual synonym map. The free function should do an atomic check as there is a potential race condition; i.e. instead of

```
--smap->ref_count;
if (smap->ref_count > 0) {
```

something like
```
if (--smap->ref_count) {
```
if you really wanted guaranteed atomicity you should use the `__sync` builtins; e.g.

```
if (__sync_sub_and_fetch(&smap->ref_count, 1)
```
see https://gcc.gnu.org/onlinedocs/gcc-4.1.0/gcc/Atomic-Builtins.html

**Comment by salt4pommes** [comment]
Will this great peace of code be part of the next release? :)

File: src/document.c [hunk 2]
```diff
@@ -135,6 +135,12 @@ RSAddDocumentCtx *NewAddDocumentCtx(IndexSpec *sp, Document *b, const char **err
     aCtx->fwIdx = NewForwardIndex(&aCtx->doc, sp->flags);
   }
 
+  if (sp->smap) {
```
Comment (dvirsky) [inline src/document.c]: Please document the rationale here, it;s not clear what happens here and why

File: src/ext/default.h [hunk 3]
```diff
@@ -2,6 +2,7 @@
 #define __EXT_DEFAULT_H__
 #include "redisearch.h"
 
+#define SYNONYMS_EXPENDER_NAME "SYN"
```
Comment (dvirsky) [inline src/ext/default.h]: The default expander should do both synonyms and stems - or rather wrap them both. 

Optionally the user can specify `EXPANDER SBSTEM|SYNONYM` to invoke just one, or `EXPANDER DEFAULT` to invoke the default. 

Please update the docs with the options.

File: src/forward_index.c [hunk 4]
```diff
@@ -216,6 +223,22 @@ int forwardIndexTokenFunc(void *ctx, const Token *tokInfo) {
...
+
+  if (tokCtx->idx->smap) {
+    TermData *t_data = SynonymMap_GetIdsBySynonym(tokCtx->idx->smap, tokInfo->tok, tokInfo->tokLen);
+    if (t_data) {
+      printf("adding sysnonym on word %.*s\r\n", tokInfo->tokLen, tokInfo->tok);
+      char synonym_buff[SYNONYM_BUFF_LEN];
+      size_t synonym_len;
+      for (int i = 0; i < array_len(t_data->ids); ++i) {
+        synonym_len = SynonymMap_IdToStr(t_data->ids[i], synonym_buff, SYNONYM_BUFF_LEN);
+        printf("synonym added %.*s\r\n", synonym_len, synonym_buff);
```
Comment (dvirsky) [inline src/forward_index.c]: remove printf

File: src/module.c [hunk 5]
```diff
@@ -1399,6 +1399,89 @@ int SuggestGetCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
   return REDISMODULE_OK;
 }
 
+int SynAddCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
```
Comment (dvirsky) [inline src/module.c]: Please document the API of the command here as well.

File: src/synonym_map.c [hunk 6]
```diff
@@ -0,0 +1,222 @@
+#include "synonym_map.h"
+#include "rmalloc.h"
+#include "util/fnv.h"
+
+#define INITIAL_CAPACITY 16
```
Comment (dvirsky) [inline src/synonym_map.c]: Change to 2

File: docs/Synonyms.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+```
+FT.SYNDUMP <index name>
+```
+Returns a list of synonym terms and their synonym group ids.
+
+## Enabling synonyms on search
+By default synonyms are not enabled when searching. In order to enable it add the EXPENDER argument to the search command with value SYN
+```
+FT.SEARCH <index> <query> EXPANDER SYN
+```
```
Comment (dvirsky) [inline docs/Synonyms.md]: Add an example in redis-cli

**Pull request merged**